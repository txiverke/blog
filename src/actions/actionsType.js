// @flow

export const LOAD_STATS_REQUEST = 'LOAD_STATS_REQUEST'
export const LOAD_STATS_SUCCESS = 'LOAD_STATS_SUCCESS'
export const LOAD_STATS_FAILURE = 'LOAD_STATS_FAILURE'

export const LOG_IN_USER_REQUEST = 'LOG_IN_USER_REQUEST'
export const LOG_IN_USER_SUCCESS = 'LOG_IN_USER_SUCCESS'
export const LOG_IN_USER_FAILURE = 'LOG_IN_USER_FAILURE'

export const IS_AUTHENTICATED_REQUEST = 'IS_AUTHENTICATE_REQUEST'
export const IS_AUTHENTICATED_SUCCESS = 'IS_AUTHENTICATE_SUCCESS'
export const IS_AUTHENTICATED_FAILURE = 'IS_AUTHENTICATE_FAILURE'

export const LOAD_USER_DATA_REQUEST = 'LOAD_USER_DATA_REQUEST'
export const LOAD_USER_DATA_SUCCESS = 'LOAD_USER_DATA_SUCCESS'
export const LOAD_USER_DATA_FAILURE = 'LOAD_USER_DATA_FAILURE'

export const UPDATE_USER_DATA_REQUEST = 'UPDATE_USER_DATA_REQUEST'
export const UPDATE_USER_DATA_SUCCESS = 'UPDATE_USER_DATA_SUCCESS'
export const UPDATE_USER_DATA_FAILURE = 'UPDATE_USER_DATA_FAILURE'

export const LOG_OUT_USER_REQUEST = 'LOG_OUT_USER_REQUEST'
export const LOG_OUT_USER_SUCCESS = 'LOG_OUT_USER_SUCCESS'
export const LOG_OUT_USER_FAILURE = 'LOG_OUT_USER_FAILURE'

export const LOAD_POST_DATA_REQUEST = 'LOAD_POST_DATA_REQUEST'
export const LOAD_POST_DATA_SUCCESS = 'LOAD_POST_DATA_SUCCESS'
export const LOAD_POST_DATA_FAILURE = 'LOAD_POST_DATA_FAILURE'

export const LOAD_POST_ITEM_REQUEST = 'LOAD_POST_ITEM_REQUEST'
export const LOAD_POST_ITEM_SUCCESS = 'LOAD_POST_ITEM_SUCCESS'
export const LOAD_POST_ITEM_FAILURE = 'LOAD_POST_ITEM_FAILURE'
export const UPDATE_POST_ITEM_SUCCESS = 'UPDATE_POST_ITEM_SUCCESS'

export const CREATE_POST_DATA_REQUEST = 'CREATE_POST_DATA_REQUEST'
export const CREATE_POST_DATA_SUCCESS = 'CREATE_POST_DATA_SUCCESS'
export const CREATE_POST_DATA_FAILURE = 'CREATE_POST_DATA_FAILURE'

export const UPDATE_POST_DATA_REQUEST = 'UPDATE_POST_DATA_REQUEST'
export const UPDATE_POST_DATA_SUCCESS = 'UPDATE_POST_DATA_SUCCESS'
export const UPDATE_POST_DATA_FAILURE = 'UPDATE_POST_DATA_FAILURE'

export const REMOVE_POST_DATA_REQUEST = 'REMOVE_POST_DATA_REQUEST'
export const REMOVE_POST_DATA_SUCCESS = 'REMOVE_POST_DATA_SUCCESS'
export const REMOVE_POST_DATA_FAILURE = 'REMOVE_POST_DATA_FAILURE'
