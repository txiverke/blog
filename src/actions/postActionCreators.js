// @flow

import * as ACTION from './actionsType'
import { setPromise } from '../utils/helpers'

const URL = 'posts'

/** LOAD POSTS */
export const loadPostDataRequest = () => ({ type: ACTION.LOAD_POST_DATA_REQUEST })
export const loadPostDataSuccess = (payload: Array<Object>) => ({ type: ACTION.LOAD_POST_DATA_SUCCESS, payload })
export const loadPostDataFailure = () => ({ type: ACTION.LOAD_POST_DATA_FAILURE })

export const loadPostData = () => 
  async (dispatch: Function) => {
    dispatch(loadPostDataRequest())
    try {
      // $FlowFixMe
      setPromise.method = 'GET'
      // $FlowFixMe
      setPromise.body = null
      // $FlowFixMe      
      setPromise.urls = URL
      const data = await setPromise.response()
      return dispatch(loadPostDataSuccess(data))
    } catch (err) {
      return dispatch(loadPostDataFailure())
    }

}

/** LOAD POST ITEM */
export const loadPostItemRequest = () => ({ type: ACTION.LOAD_POST_ITEM_REQUEST })
export const loadPostItemSuccess = (payload: Object) => ({ type: ACTION.LOAD_POST_ITEM_SUCCESS, payload })
export const loadPostItemFailure = () => ({ type: ACTION.LOAD_POST_ITEM_FAILURE })
export const updatePostItemSucces = () => ({ type: ACTION.UPDATE_POST_ITEM_SUCCESS })

export const loadPostItem = (id: string) => 
  async (dispatch: Function) => {
    dispatch(loadPostItemRequest())

    try {
      // $FlowFixMe      
      setPromise.method = 'GET'
      // $FlowFixMe      
      setPromise.body = null
      // $FlowFixMe      
      setPromise.urls = `${URL}/${id}`
      const data = await setPromise.response()
      return dispatch(loadPostItemSuccess(data))
    } catch (err) {
      return dispatch(loadPostItemFailure())
    }

}

/** CREATE POST */
export const createPostDataRequest = () => ({ type: ACTION.CREATE_POST_DATA_REQUEST })
export const createPostDataSuccess = (payload: Array<Object>) => ({ type: ACTION.CREATE_POST_DATA_SUCCESS, payload })
export const createPostDataFailure = () => ({ type: ACTION.CREATE_POST_DATA_FAILURE})

export const createPostData = (obj: Object) => 
  async (dispatch: Function) => {
    dispatch(createPostDataRequest())
    
    try {
      const body = new FormData()
      body.append('file', obj.file, obj.file.name);
      body.append('title', obj.title)
      body.append('content', obj.content)
      body.append('tags', obj.tags)
      body.append('link', obj.link)

      // $FlowFixMe      
      setPromise.method = 'POST'
      // $FlowFixMe
      setPromise.body = body
      // $FlowFixMe
      setPromise.urls = URL

      const data = await setPromise.response()
      return dispatch(createPostDataSuccess(data))
    } catch (err) {
      return dispatch(createPostDataFailure())
    }
}

/** UPDATE POST */
export const updatePostDataRequest = () => ({ type: ACTION.UPDATE_POST_DATA_REQUEST })
export const updatePostDataSuccess = (payload: Array<Object>) => ({ type: ACTION.UPDATE_POST_DATA_SUCCESS, payload })
export const updatePostDataFailure = () => ({ type: ACTION.UPDATE_POST_DATA_FAILURE})

export const updatePostData = (obj: Object, id: string) => 
  async (dispatch: Function) => {
    dispatch(updatePostDataRequest())

    try {
      let body = new FormData()
      if (obj.file) body.append('file', obj.file, obj.file.name)
      body.append('title', obj.title)
      body.append('content', obj.content)
      body.append('tags', obj.tags)
      body.append('link', obj.link)

      // $FlowFixMe
      setPromise.method = 'PUT'
      // $FlowFixMe
      setPromise.body = body
      // $FlowFixMe
      setPromise.urls = `${URL}/${id}`
      const data = await setPromise.response()
      dispatch(updatePostDataSuccess(data))
      return dispatch(updatePostItemSucces())
    } catch (err) {
      return dispatch(updatePostDataFailure())
    }
}

/** REMOVE POST */
export const removePostDataRequest = () => ({ type: ACTION.REMOVE_POST_DATA_REQUEST })
export const removePostDataSuccess = (payload: Data) => ({ type: ACTION.REMOVE_POST_DATA_SUCCESS, payload })
export const removePostDataFailure = () => ({ type: ACTION.REMOVE_POST_DATA_FAILURE})

export const removePostData = (id: string) => 
  async (dispatch: Function) => {
  dispatch(removePostDataRequest())

  try {
    // $FlowFixMe
    setPromise.method = 'DELETE'
    // $FlowFixMe
    setPromise.body = null
    // $FlowFixMe
    setPromise.urls = `${URL}/${id}`
    const data = await setPromise.response()
    return dispatch(removePostDataSuccess(data))
  } catch (err) {
    return dispatch(removePostDataFailure())
  }
}

/** CHECK TAGS */

export const checkPostsTagsRequest = () => ({ type: ACTION.CHECK_POSTS_TAGS_REQUEST })
export const checkPostsTagsSuccess = (payload: Array<string>) => ({ type: ACTION.CHECK_POSTS_TAGS_SUCCESS, payload })
export const checkPostsTagsFailure = () => ({ type: ACTION.CHECK_POSTS_TAGS_FAILURE })

export const checkPostsTags = (arr: Array<string>) => 
  (dispatch: Function) => {
    dispatch(checkPostsTagsRequest())

    if (Array.isArray(arr)) {
      return dispatch(checkPostsTagsSuccess(arr))
    } 

    return dispatch(checkPostsTagsFailure())
  }