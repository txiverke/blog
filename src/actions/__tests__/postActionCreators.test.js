// @flow

import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import * as action from '../postActionCreators'
import * as type from '../actionsType'

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

describe('postActionCreators', () => {
  describe('load post data actions', () => {
    it('should return LOAD_POST_DATA_REQUEST action', () => {
      expect(action.loadPostDataRequest()).toMatchSnapshot()
    })

    it('should return LOAD_POST_DATA_SUCCESS action', () => {
      const payload = [{ test: true }]
      expect(action.loadPostDataSuccess(payload)).toMatchSnapshot()
    })

    it('should return LOAD_POST_DATA_FAILURE', () => {
      expect(action.loadPostDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})

      return store.dispatch(action.loadPostData())
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toMatchSnapshot()
       })
    })
  })

  describe('load post item actions', () => {
    it('should return LOAD_POST_ITEM_REQUEST action', () => {
      expect(action.loadPostItemRequest()).toMatchSnapshot()
    })

    it('should return LOAD_POST_ITEM_SUCCESS action', () => {
      const payload = { test: true }
      expect(action.loadPostItemSuccess(payload)).toMatchSnapshot()
    })

    it('should return LOAD_POST_ITEM_FAILURE', () => {
      expect(action.loadPostItemFailure()).toMatchSnapshot()
    })

    it('should return UPDATE_POST_ITEM_SUCCESS action', () => {
      expect(action.updatePostItemSucces()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})

      return store.dispatch(action.loadPostItem(''))
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toMatchSnapshot()
       })
    })
  })

  describe('create post data actions', () => {
    it('should return CREATE_POST_DATA_REQUEST action', () => {
      expect(action.createPostDataRequest()).toMatchSnapshot()
    })

    it('should return CREATE_POST_DATA_SUCCESS action', () => {
      const payload = [{ test: true }]
      expect(action.createPostDataSuccess(payload)).toMatchSnapshot()
    })

    it('should return CREATE_POST_DATA_FAILURE', () => {
      expect(action.createPostDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})

      return store.dispatch(action.loadPostData({}))
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toMatchSnapshot()
       })
    })
  })

  describe('update post data actions', () => {
    it('should return UPDATE_POST_DATA_REQUEST action', () => {
      expect(action.updatePostDataRequest()).toMatchSnapshot()
    })

    it('should return UPDATE_POST_DATA_SUCCESS action', () => {
      const payload = [{ test: true }]
      expect(action.updatePostDataSuccess(payload)).toMatchSnapshot()
    })

    it('should return UPDATE_POST_DATA_FAILURE', () => {
      expect(action.updatePostDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})

      return store.dispatch(action.updatePostData({}, ''))
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toMatchSnapshot()
       })
    })
  })

  describe('remove post data actions', () => {
    it('should return REMOVE_POST_DATA_REQUEST action', () => {
      expect(action.removePostDataRequest()).toMatchSnapshot()
    })

    it('should return REMOVE_POST_DATA_SUCCESS action', () => {
      const payload = {
        completed: true,
        data: [],
        message: 'fake message',
        error: false
      }
      expect(action.removePostDataSuccess(payload)).toMatchSnapshot()
    })

    it('should return REMOVE_POST_DATA_FAILURE', () => {
      expect(action.removePostDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})

      return store.dispatch(action.removePostData(''))
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toMatchSnapshot()
       })
    })
  })

  describe('check post tags actions', () => {
    it('should return CHECK_POSTS_TAGS_REQUEST action', () => {
      expect(action.checkPostsTagsRequest()).toMatchSnapshot()
    })

    it('should return CHECK_POSTS_TAGS_SUCCESS action', () => {
      const payload = ['test']
      expect(action.checkPostsTagsSuccess(payload)).toMatchSnapshot()
    })

    it('should return CHECK_POSTS_TAGS_FAILURE', () => {
      expect(action.checkPostsTagsFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on successful request', () => {
      const store = mockStore({})
      const expectedActions = [
        'CHECK_POSTS_TAGS_REQUEST',
        'CHECK_POSTS_TAGS_SUCCESS'
      ]
      store.dispatch(action.checkPostsTags(['tests']))
        
      const actualActions = store.getActions().map(action => action.type)
      expect(actualActions).toEqual(expectedActions)
    })
  })
  
})