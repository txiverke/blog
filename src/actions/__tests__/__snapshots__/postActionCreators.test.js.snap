// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postActionCreators check post tags actions should return CHECK_POSTS_TAGS_FAILURE 1`] = `
Object {
  "type": "CHECK_POSTS_TAGS_FAILURE",
}
`;

exports[`postActionCreators check post tags actions should return CHECK_POSTS_TAGS_REQUEST action 1`] = `
Object {
  "type": "CHECK_POSTS_TAGS_REQUEST",
}
`;

exports[`postActionCreators check post tags actions should return CHECK_POSTS_TAGS_SUCCESS action 1`] = `
Object {
  "payload": Array [
    "test",
  ],
  "type": "CHECK_POSTS_TAGS_SUCCESS",
}
`;

exports[`postActionCreators create post data actions expected actions should be dispatched on request 1`] = `
Array [
  "LOAD_POST_DATA_REQUEST",
  "LOAD_POST_DATA_FAILURE",
]
`;

exports[`postActionCreators create post data actions should return CREATE_POST_DATA_FAILURE 1`] = `
Object {
  "type": "CREATE_POST_DATA_FAILURE",
}
`;

exports[`postActionCreators create post data actions should return CREATE_POST_DATA_REQUEST action 1`] = `
Object {
  "type": "CREATE_POST_DATA_REQUEST",
}
`;

exports[`postActionCreators create post data actions should return CREATE_POST_DATA_SUCCESS action 1`] = `
Object {
  "payload": Array [
    Object {
      "test": true,
    },
  ],
  "type": "CREATE_POST_DATA_SUCCESS",
}
`;

exports[`postActionCreators load post data actions expected actions should be dispatched on request 1`] = `
Array [
  "LOAD_POST_DATA_REQUEST",
  "LOAD_POST_DATA_FAILURE",
]
`;

exports[`postActionCreators load post data actions should return LOAD_POST_DATA_FAILURE 1`] = `
Object {
  "type": "LOAD_POST_DATA_FAILURE",
}
`;

exports[`postActionCreators load post data actions should return LOAD_POST_DATA_REQUEST action 1`] = `
Object {
  "type": "LOAD_POST_DATA_REQUEST",
}
`;

exports[`postActionCreators load post data actions should return LOAD_POST_DATA_SUCCESS action 1`] = `
Object {
  "payload": Array [
    Object {
      "test": true,
    },
  ],
  "type": "LOAD_POST_DATA_SUCCESS",
}
`;

exports[`postActionCreators load post item actions expected actions should be dispatched on request 1`] = `
Array [
  "LOAD_POST_ITEM_REQUEST",
  "LOAD_POST_ITEM_FAILURE",
]
`;

exports[`postActionCreators load post item actions should return LOAD_POST_ITEM_FAILURE 1`] = `
Object {
  "type": "LOAD_POST_ITEM_FAILURE",
}
`;

exports[`postActionCreators load post item actions should return LOAD_POST_ITEM_REQUEST action 1`] = `
Object {
  "type": "LOAD_POST_ITEM_REQUEST",
}
`;

exports[`postActionCreators load post item actions should return LOAD_POST_ITEM_SUCCESS action 1`] = `
Object {
  "payload": Object {
    "test": true,
  },
  "type": "LOAD_POST_ITEM_SUCCESS",
}
`;

exports[`postActionCreators load post item actions should return UPDATE_POST_ITEM_SUCCESS action 1`] = `
Object {
  "type": "UPDATE_POST_ITEM_SUCCESS",
}
`;

exports[`postActionCreators remove post data actions expected actions should be dispatched on request 1`] = `
Array [
  "REMOVE_POST_DATA_REQUEST",
  "REMOVE_POST_DATA_FAILURE",
]
`;

exports[`postActionCreators remove post data actions should return REMOVE_POST_DATA_FAILURE 1`] = `
Object {
  "type": "REMOVE_POST_DATA_FAILURE",
}
`;

exports[`postActionCreators remove post data actions should return REMOVE_POST_DATA_REQUEST action 1`] = `
Object {
  "type": "REMOVE_POST_DATA_REQUEST",
}
`;

exports[`postActionCreators remove post data actions should return REMOVE_POST_DATA_SUCCESS action 1`] = `
Object {
  "payload": Object {
    "completed": true,
    "data": Array [],
    "error": false,
    "message": "fake message",
  },
  "type": "REMOVE_POST_DATA_SUCCESS",
}
`;

exports[`postActionCreators update post data actions expected actions should be dispatched on request 1`] = `
Array [
  "UPDATE_POST_DATA_REQUEST",
  "UPDATE_POST_DATA_FAILURE",
]
`;

exports[`postActionCreators update post data actions should return UPDATE_POST_DATA_FAILURE 1`] = `
Object {
  "type": "UPDATE_POST_DATA_FAILURE",
}
`;

exports[`postActionCreators update post data actions should return UPDATE_POST_DATA_REQUEST action 1`] = `
Object {
  "type": "UPDATE_POST_DATA_REQUEST",
}
`;

exports[`postActionCreators update post data actions should return UPDATE_POST_DATA_SUCCESS action 1`] = `
Object {
  "payload": Array [
    Object {
      "test": true,
    },
  ],
  "type": "UPDATE_POST_DATA_SUCCESS",
}
`;
